invenio_preservation_sync/receivers.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 CERN.
+#
+# Invenio-Preservation-Sync is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Receiver for managing Preservation Sync events integration."""
+
+from flask import current_app, g
+from invenio_webhooks.models import Receiver
+from werkzeug.exceptions import BadRequest
+
+from invenio_preservation_sync.api import PreservationInfoAPI
+from invenio_preservation_sync.utils import get_mandatory_config
+
+from .errors import (
+    InvalidSenderError,
+    MissingConfigError,
+    PreservationAlreadyReceivedError,
+)
+
+
+class PreservationSyncReceiver(Receiver):
+    """Handle incoming notification from GitHub on a new release."""
+
+    def run(self, event):
+        """Process an event."""
+        try:
+            perm_policy = get_mandatory_config(
+                current_app, "PRESERVATION_SYNC_PERMISSION_POLICY"
+            )
+            pid_resolver = get_mandatory_config(
+                current_app, "PRESERVATION_SYNC_PID_RESOLVER"
+            )
+
+            pid_id = event.payload.get("pid")
+            revision_id = event.payload.get("revision_id")
+            status = event.payload.get("status")
+
+            if not pid_id or not revision_id or not status:
+                raise BadRequest("Mandatory fields are missing from the request.")
```

Comment:
I like that you’re returning an HTTP error, which is appropriate since this is being received by a webhook.

However, your receiver is currently handling different types of errors:
- **HTTP Errors:** This is good practice since the receiver acts as an HTTP interface for your system.
- **Business Logic Errors**: It’s less ideal to handle these at this level. Business logic errors, like permission errors, should be raised by the business layer and handled by the interface or presentation layer. Each interface might handle a permission error differently—for example, an HTTP API would return a 403 Permission Denied, while a Celery task might handle the permission error in another way.


For reference, see how we handle these exceptions globally and add consistent error codes: https://github.com/alejandromumo/flask-resources/blob/master/flask_resources/errors.py